jets_bench/benches/elements/main.rs

```diff
@@ -0,0 +1,643 @@
+        (Elements::GejRescale, InputSampling::Custom(Arc::new(gej_fe_pair))),
+        (Elements::GejNormalize, InputSampling::Custom(Arc::new(|| SimplicityGej::sample().value()))),
+        (Elements::GejNegate, InputSampling::Custom(Arc::new(|| SimplicityGej::sample().value()))),
+        (Elements::GejDouble, InputSampling::Custom(Arc::new(|| SimplicityGej::sample().value()))),
+        (Elements::GejAdd, InputSampling::Custom(Arc::new(gej_pair))),
+        (Elements::GejGeAddEx, InputSampling::Custom(Arc::new(gej_ge_pair))),
+        (Elements::GejGeAdd, InputSampling::Custom(Arc::new(gej_ge_pair))),
+        (Elements::GejIsInfinity, InputSampling::Custom(Arc::new(|| SimplicityGej{ ge: SimplicityGe::sample(), z: SimplicityFe([0u8; 32]) }.value()))),
+        (Elements::GejXEquiv, InputSampling::Custom(Arc::new(fe_gej_pair))),
+        (Elements::GejYIsOdd, InputSampling::Custom(Arc::new(|| SimplicityGej::sample().value()))),
+        (Elements::GejIsOnCurve, InputSampling::Custom(Arc::new(|| SimplicityGej::sample().value()))),
+        // Other jets
+        (Elements::GeIsOnCurve, InputSampling::Custom(Arc::new(|| SimplicityGe::sample().value()))),
+        (Elements::Scale, InputSampling::Custom(Arc::new(scalar_gej_pair))),
+        (Elements::Generate, InputSampling::Custom(Arc::new(|| SimplicityScalar::sample().value()))),
+        (Elements::LinearCombination1, InputSampling::Custom(Arc::new(scalar_gej_scalar_pair))),
+        (Elements::LinearVerify1, InputSampling::Custom(Arc::new(linear_verify))),
+        (Elements::Decompress, InputSampling::Custom(Arc::new(|| SimplicityPoint::sample().value()))),
+        (Elements::PointVerify1, InputSampling::Custom(Arc::new(point_verify))),
+        // Signature jets
+        (Elements::Bip0340Verify, InputSampling::Custom(Arc::new(bip_0340))),
+        (Elements::CheckSigVerify, InputSampling::Custom(Arc::new(check_sig_verify))),
+    ];
+    for (jet, sample) in arr {
+        let src_ty = jet.source_ty().to_type(&pows_of_two);
+        let tgt_ty = jet.target_ty().to_type(&pows_of_two);
+
+        let mut group = c.benchmark_group(&jet.to_string());
+        let env = EnvSampling::Null.env();
+
+        for i in 0..NUM_RANDOM_SAMPLES {
+            let params = JetParams::with_rand_aligns(sample.clone());
+            // Assumption: `buffer.write` is non-negligible
+            let bench_name = format!("{}", i);
+            group.bench_with_input(bench_name, &params,|b, params| {
+                b.iter_batched(
+                    || {
+                        let mut buffer = JetBuffer::new(&src_ty, &tgt_ty, params);
+                        let (src, dst) = buffer.write(&src_ty, params, &mut rng);
+                        (dst, src, &env, buffer)
+                    },
+                    |(mut dst, src, env, _buffer)| jet.c_jet_ptr()(&mut dst, src, env.c_tx_env()),
+                    BatchSize::SmallInput,
+                )
+            });
+        }
```

Comment:
This creates a sub-benchmark for every separate input we test. Is that what we want? An obvious upside is that we can use `bench_with_input` (which is canonical, but I haven't yet understood when it should be used). A downside is that there is a 3 second warmup for every sub-benchmark (which might improve accuracy, but is it really needed?) and the resulting criterion data is slightly harder to process in Python (maybe cargo-criterion can compute percentiles). Depending on what we want, 50 samples might be too few for a meaningful distribution.

The alternative would be to write a custom time loop which is called multiple times by criterion, and which makes sure that the same input is used for each sample. Obvious downside is that this is not canonical criterion use.

Reply:
There was another push. I see now that we do a two-step approach: First we run sub-benchmarks for `n` inputs that we think are worst-case. Then, we run the `main.rs` script to find the worst among them.

Reply:
I think even 10 samples should be good enough. The distributions are selected from white-listing what we consider to be the worst case input. I would even argue that even 1/2 is enough, but that might be off because some of the jets do actually randomly select envs.
 
Regarding a custom timing loop, I am not against it. This is just meant as a starter that we have something benchmarks ongoing while we discuss specifics. My intuition is that it would not matter, so feel free to have a go at this with a custom loop. I think would make things harder to review because we are calculating our own times, starting and stoping clocks, etc. 

Regarding warm up time etc, time per loop etc. All of this is configurable. Ultimately, we are doing the same calculation in the custom time loop vs here. Just that the default parameters here make it seem like taking longer(which might be good). Something like benchmark `measurement_time ` 100ms with 500 num samples would be same as one would do in custom loop.

Reply:
> There was another push. I see now that we do a two-step approach: First we run sub-benchmarks for n inputs that we think are worst-case. Then, we run the main.rs script to find the worst among them.

Yep, this was what I always intended to do. This is done in last commit. (modulo fixup commits)